catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,445 @@
+   * #setFederatedSources(List)}, {@link #setCatalogProviders(List)}, and {@link
+   * #setCatalogStores(List)} are discovered.
+   *
+   * <p>A process to check (or recheck) the availability of each {@link Source} is started unless
+   * the {@link Source} was created or modified within the last poll interval and the availability
+   * check triggered by that action (see {@link #event(ServiceEvent, Map)} has not yet completed.
+   *
+   * <p>TODO Should this method be synchronized?
+   */
+  private void handleAllSources() {
+    final Set<Source> sources =
+        Stream.of(connectedSources, federatedSources, catalogProviders, catalogStores)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toSet());
+
+    LOGGER.trace("Found {} sources", sources.size());
+
+    // Unbind orphaned entries in the map caused by when sources are modified and their key is
+    // different
+    final Map<SourceKey, Source> sourceKeyMap =
+        sources
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    SourceKey::new,
+                    source -> source,
+                    (source, anotherSourceWithTheSameSourceKey) -> {
+                      LOGGER.warn(
+                          "Found different sources with the same SourceKey with id={}. This means that the SourcePoller might not report accurate availability for these sources in the CatalogFramework. Confirm that every source has a unique id, and try restarting.",
+                          source.getId());
+                      return source;
+                    }));
+
+    final Set<SourceKey> currentSourceKeys = sourceKeyMap.keySet();
+    final Set<SourceKey> sourceKeysInTheCache = cache.asMap().keySet();
+
+    final Set<SourceKey> sourceKeysToRemove = new HashSet<>(sourceKeysInTheCache);
+    sourceKeysToRemove.removeAll(currentSourceKeys);
+    if (CollectionUtils.isNotEmpty(sourceKeysToRemove)) {
+      LOGGER.debug(
+          "Found {} SourceKeys to remove from the cache. This is caused by when sources are destroyed or modified in a way that their SourceKey changes. Removing these from the cache now",
+          sourceKeysToRemove.size());
+      sourceKeysToRemove.forEach(cache::invalidate);
+    }
+
+    for (final Map.Entry<SourceKey, Source> entry : sourceKeyMap.entrySet()) {
+      final Source source = entry.getValue();
+      final SourceKey sourceKey = entry.getKey();
+
+      final Optional<CachedSourceAvailability> cachedSourceAvailabilityOptional =
+          Optional.ofNullable(cache.getIfPresent(sourceKey));
+      if (cachedSourceAvailabilityOptional.isPresent()) {
+        final CachedSourceAvailability cachedSourceAvailability =
+            cachedSourceAvailabilityOptional.get();
+        LOGGER.trace(
+            "Starting a process to recheck the availability of source id={}, which has already been added to the cache ",
+            source.getId());
```

Comment:
Reminder to remove this log statement. It duplicates a better message that is logged in the `recheckAvailability` method.